File: 402.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if num.count('0') + k >= len(num):
            return '0'
        st = []
        delCount = 0
        for c in num:
            while delCount < k and len(st) > 0 and c < st[-1]:
                st.pop()
                delCount += 1
            st.append(c)
        for _ in range(k - delCount):
            st.pop()
        res = ''.join(st).lstrip('0')
        if len(res) == 0:
            return '0'
        else:
            return res

File: sol/402/402.cpp
struct Solution {
    string removeKdigits(string &num, int k) {
        int zeroCount = 0;
        for (char &c : num){
            zeroCount += c == '0';
        }
        if (zeroCount + k >= num.size()){
            return "0";
        }
        string res;
        for (char &c : num){
            while (k > 0 && res.size() > 0 && c < res.back()){
                res.pop_back();
                --k;
            }
            if (res.size() > 0 || c != '0'){
                res.push_back(c);
            }
        }
        while (res.size() > 0 && k-- > 0){
            res.pop_back();
        }
        if (res.size() == 0){
            return "0";
        } else {
            return res;
        }
    }
};
